### ISSUE TITLE
add scripting language.

### ISSUE THREAD

**Issue by bushig** [kremius/karya-valya#170]
C++ is performant, but game logic should be coded in some scripting language. For example Python, JavaScript or Lua. Hope it won't be custom lang.

**Comment by kremius** [comment]
Scripting language for game logic is a tempting idea, but it entails a lot of consequences.

Pros:
* It will be easier to implement new game stuff for some coders

Cons:
* Scripts are slow
* Complicated and time-consuming issue

So it is pretty questionable. I will think about it if there will not be any big issues left, but I don't think it will really happen.

**Comment by bushig** [comment]
You also should keep i mind that its not cool to recompile game every time you change something in game logic. And scripting language don't really have to be fast, becouse performance dependent things like atmos should be written in C++. Main purpose of scripting language is to make game modification easy and possiple at runtime. 
You might want to take look at something like [PySDL2] (http://pysdl2.readthedocs.org/en/latest/integration.html) which is quite performant. I have also heard that lua is easier to implement.

**Comment by kremius** [comment]
I don't think that you understand the issue.

**Comment by PJB3005** [comment]
Just want to add to this:

One of the primary reasons SS13 (Cataclysm DDA is an amazing example too) has so much content is because of how easy it is to code for. 

C++ is a very hard language for beginners,  and will scare a lot of people off. 

Though yes I can definitely see how hard it is to implement.

**Comment by bauen1** [comment]
👍 
I would suggest Lua for being lightweight.

**Comment by tuesdayMarch** [comment]
What about a transpiler? 

I'm not acquainted with DM very well, but looking at TG source-code it doesn't seem very complex.

**Comment by bauen1** [comment]
Everything but DM, DM is horrible.

**Comment by tuesdayMarch** [comment]
At the very least, a transpiler would incentivize adoption for large codebases.

**Comment by bauen1** [comment]
The biggest reason for a rewrite is performance, and DM is a mess in both code and performance

**Comment by tuesdayMarch** [comment]
I don't think we're on the same page.

A transpiler would take DM shitcode and translate it into C++. Then, the codebases can still maintain their servers, but also allow their maintainers to tinker with C++ and, hopefully, shift over to C++ fully in due time.

**Comment by PJB3005** [comment]
Transpilers don't work because BYOND doesn't work.

If you want a transpiler to work you have to provide an API like BYOND. BYOND's API is a giant pile of trash so that's just infeasible.

**Comment by tuesdayMarch** [comment]
Of course it wouldn't be straight translation, it wouldn't work at all.

If the engine was divided into blank "templates," the transpiler could fill in those templates with each codebase's take.

**Comment by dangranos** [comment]
Sorry to interrupt you, but that was already done. Of course, it used C#,
but it was transpiling DM into C#. There is even a post by author about why
it was horrible and awful idea in first place.

Read this
https://www.reddit.com/r/SS13/comments/4toxuq/sooooo_what_happened_to_somnium13/

Now, being back in topic of scripting. What this project needs, is to move
the content into some external definition **right now**. By this, I mean
that you don't fucking hardcode every item, glass and so on. Engine itself
should contain no comment, only provide base to load it from definition
files. Scripting, or at least external modules,  with moving content out of
engine, are a must and I would say biggest priority for the project right
now. If this won't be done any time soon, the whole thing will become a
monolithic mess, that nobody will be able to modify without huge headaches.

**Comment by tuesdayMarch** [comment]
It looks like the author transpiled the Byond engine, which I believe was the core problem after doing a quick look-over. Nonetheless, I won't fill up this issue and work on it as a personal project.

I agree with the second point. If this were to get script extensibility, it must be divided into modules and module processor. 

However, if every module were to get scripting support it would fall back into the issue Byond has: it's too slow and bloated. You cannot let the engine, and other resource-heavy features, be run on your run of the mill scripts. 

Lua would be too slow for this. If it were to use any, my money would be on the D programming language. It's easy enough to script in, but can still get close to native C++ performance.

But we need to come back to:

- Complicated and time-consuming issue

Is it worth it? Taking a few months out of the roadmap to make it easier for people to continue writing shit code?

**Comment by bauen1** [comment]
Lua is easy to learn and easy to code.
Yes its slow, but you don't have to / shouldn't implement things like the atmo engine in lua.
If lua really gets too slow there is LuaJIT which improves performance a lot.
(http://luajit.org/performance_x86.html)

**Comment by PJB3005** [comment]
It also is hell to maintain in any form and even DM would be better, so it's off the table.

**Comment by dangranos** [comment]
Or how about: divide external things into code modules, that give you
components or functions to be assigned to objects, and object definition
files in something like json or yaml? That way, the code is kept fast, and
it will be easy enough to modify without messing with the engine itself.
And, can I ask you, where the f-- hell did you got this "Lua is hard to
maintain"? And, comparing it with DM is just weird. DM is glorified thing
for defining variables of items, that somehow got scripting added to it.
And Lua is a language, programming language, designed to be gay and light,
to be used in *embedded devices*. Which obviously means several MEGAbytes
of RAM and pretty tiny CPU, compared to ones in desktops and actual
dedicated servers that should host things.

> Is it worth it? Taking a few months out of the roadmap to make it easier
for people to continue writing shit code?

It depends on how that will be implemented, really. If it will force you to
basically dance around the hardcoded items and things of the engine, no, it
won't be.
But if it will let you, for example, add a random super drink into the bar
without recompiling the whole engine because you had to add a new fluid,
that would be awesome and very good thing to spend time on. This was
example of making some json/yaml/XML/whatever definition files for liquids
and for consumables that will just let one set colour, sprite, recipe of
it, and what functions to use when dealing with clicks and etc. One piece
of scripted code could be reused on few items, or dozens of them, that do
similar things. Like pipes, drinks, food, guns, projectiles, so on.

> However, if every module were to get scripting support it would fall back
into the issue Byond has: it's too slow and bloated. You cannot let the
engine, and other resource-heavy features, be run on your run of the mill
scripts.

The problem of BYOND was that, while providing the engine for DM, it did
so, well, see comment about variables thing. The whole thing is almost two
decades out of date, new features were added on top of old ones, if added
at all.

Also. No, somnium was no transpiling BYOND, it was transpiling compiled DM
code. You know, those binary .dmb files that you put into Dream Daemon. And
you literally suggested to do exactly what somnium did, with C# replaced by
C++. You did read his post, and that piece of code he threw? What he
basically ended up doing was reimplementing the while BYOND engine /and/
make transpiler of its code into C#. And you suggested to do that.

And wow, that was long. Hope not too long to read it all.

**Comment by kremius** [comment]
@tuesdayMarch
Transpiler is a very tempting idea on a first glance (poof - and we have a lot of gameplay and the game is ready to play), but some deep issues manifest itself after some consideration.
1. Byond API is big and not simple. It should be reproduced exactly same or there will be a lot of errors.
2. DM has a very forgiving runtime. A byond server can spam a tons of critical errors to log and still be able to work! C++ (or any other language) is not so forgiving, and in case of such errors it will probably crash. So even API will be reproduced perfectly then the final result **STILL** will be not playable because almost every build has somethere inside poorly coded parts which cause runtime errors.
3. Dynamic typing. Something like `/proc/foo(var/a, var/b)` is totally allowed in DM. The `:` operator does not help either (`something:callback()`). Sure, it is possible to somehow imitate all that stuff, but it starts to look like bad copy of the byond runtime.
4. `sleep` and `spawn`. Almost impossible to implement such async-like features in C++.
5. Much more other will appear during the process.

Overall, it will request an immense amount work - and we get a poorly looking and poorly coded mess with old and freshly introduced bugs. Probably it will be just easier to write the game from scratch.

@dangranos
Monolithic mess and scripting/external modules are quite orthogonal to each other. An engine can have scripts with an unbearable hardcoded mess in them, and engine can be carefully coded as a single executable with well separated parts.

The language choice for scripting is not really matter that much. It is not hard to attach some script engine to the game - most of scripting languages were designed with that in mind and have really easy ways to embed them into a project. The most complicated part is to provide stable API to a scripting language.
Let's assume that there is scripting language in Griefly right now and it is used for gameplay stuff. The first thing ever what will happen - I will change something in the code and API will be broken - and it will happen like almost every commit. Most scripts engines do not compile scripts, so the API mismatch will be discovered who-knows-when.
Moreover, functionality of Griefly is lacking now (e.g. you cannot make proper GUI for a terminal now) and so noone will use scripting when it is impossible to script anything beside basic stuff - the engine is not ready yet.

So, in order to implement scripting the stable API is needed. In order to get stable API some subset of SS13 should be implemented on Griefly. Such subset will show 1) API is quite stable 2) API is ready for use. The easiest way is to code this subset in C++ - that way most changes in API are solved with an autoreplacement, or in the worst case a compile error will be shown.

But since the code is open source and there will be SS13 in miniature - maybe it will be just easier to code everything in C++? Sure, some people will be scared away by C++, but others will come. It is an open question for now, but for sure it can be postponed until stable API will be ready.

Btw, just changing some variables without code required (like various types of clothes) is indeed a nice feature (and relatively easy to implement), so it is planned in a near future #280.

**Comment by shelltitan** [comment]
I think a proper documentation would be better than a scripting language at the moment.

**Comment by tuesdayMarch** [comment]
@shelltitan 
Git tracks commits so it shouldn't be too hard to keep it updated once a base doc is in place.

I'm brushing up on C++ and Go, so I'll draft something up soon in English.

**Comment by dangranos** [comment]
That Go thing really should be replaced by a built-in functionally of
"server" in the main application. Because honestly, it seems that nobody
knows how to maintain it except for one guy over there. That might or might
not be able to modify it in the future. The whole thing is very simple to
integrate into C++ engine before it's too late.

**Comment by tuesdayMarch** [comment]
@dangranos 

Can you elaborate on what you mean by:

>That Go thing really should be replaced by a built-in functionally of "server" in the main application.

**Comment by shelltitan** [comment]
The server doesn't do simulation just forwards your message and I think check connections. If you want to make a server that simulates then it might be worth to rewrite.

**Comment by dangranos** [comment]
@shelltitan That's just duplication of features. What I meant, is that the functionality of authorization and relaying **should** be in the griefly application itself. Completely. The Go """server""" should not be a thing. It's just weird to have a separate program, in different language, that requires another few compilation steps, requires twice as much effort as launching a game, all for no benefits at all. It only makes it harder to start up and maintain the project.

I know that server does none of simulations, because griefly is lockstep at its core and I don't see this being changed any time soon, despite any claims made in the past about whole "rework networking protocol".

**Comment by kremius** [comment]
Please, consider to move your discussion about to appropriate thread - #375 
All network issues were covered quite in details there, and if you are still have some questions you can ask them in the thread.

**Comment by minexew** [comment]
V8 is anything but slow. Huge, yes. Cumbersome to build, certainly. But it's blazing fast.

Going with Javascript seems to me like the best route by far, because it also translates trivially into browsers (of course, you would need to have the runtime there as well, which is definitely a non-trivial amount of effort -- but emscripten is a thing now, and it will shine even more when wasm rolls out universally)

Lua's main advantage is that it's small. That advantage diminishes as the project written in it grows -- in fact, it becomes a disadvantage pretty fast. Look at some of the most complex WoW addons. it's a mess.

Python is a great language, but none of the implementations is really suitable for real-time networked games. This is not just about interpretation speed, but also the GC and memory management in general.

Either way, by now it's pretty clear that Griefly doesn't have a shot at being successful without providing a fantastic scripting experience. If it does, and if we can figure out a way to port the huge existing body of DM code gradually, it can definitely be the future of SS13.

**Comment by PJB3005** [comment]
My personal suggestion nowadays is C# through embedded Mono.

1. Fast.
2. Common in the industry (Unity, Monogame)
3. Amazing IDE support.
4. Scales to projects the scale of SS13.

**Comment by kremius** [comment]
Worth to mention that current scripting concept assumes an allowed reentrancy in a script runtime (script function call -> native function call -> script function call). Regular .NET runtime does not support such feature (reentrancy MDA). I cannot find similar info about Mono & V8, but wasm reentrability hints that the feature can be present in V8 engine.
Personally I would prefer a strongly typed language (such as C#), but in the end even Javascript can be made strongly typed with help of some transpilers (like Typescript).

**Comment by PJB3005** [comment]
I'm not sure what you're talking about with the reentrancy MDA thing, because that only seems to apply when you've *incorrectly* jumped between managed and native stack. You can totally export function "pointers" native can call to invoke .NET functions. See [SFML.NET's Stream Adaptor](https://github.com/SFML/SFML.Net/blob/master/src/Window/StreamAdaptor.cs#L14)

**Comment by minexew** [comment]
Mono is great, but mandatory typing still adds friction when fast prototyping. Game logic tends to be very dynamic in typing (large hierarchies of classes derived from Actor/Entity), so C# doesn't seem quite like the best match to me. (OTOH, it's a great language for implementing a game *engine*) Static checking and IDE support are definite pluses though.

TypeScript could be a good compromise. Re-entrancy shouldn't be a problem in either Mono or V8.

**Comment by kremius** [comment]
@PJB3005 It is not about calling once (that's possible for sure), it is about chain of calls where you enter and then **re**enter runtime in the same call stack. E.g. in the some game object which is written in C# call to a native function is made. The native function calls (maybe indirectly) some method from some game object which is written in C# too -> re-entrance happened.

**Comment by PJB3005** [comment]
@minexew This is SS13. You either have static typing or you don't, and not having it on a project the complexity of SS13 is not worth the marginally quicker prototyping.

@kremius SFML.NET does that with its stream adaptors, and Godot does it with their Mono bindings. I don't think that's a problem.

Native C++ has streams for say file loading. SFML.NET makes a dummy type to bind any .NET `Stream` to this by exposing function pointers the C++ code calls. It goes like this: C# tells SFML.NET to load from say a zip stream, SFML.NET wraps the stream and handles it to the C++ code, C++ code then invokes the read methods on that wrapper, which execute the stream's managed code to read the actual bytes.

**Comment by kremius** [comment]
@PJB3005 Sounds pretty good then.

**Comment by minexew** [comment]
@PJB3005 Given the amount of work you've done on one of the most prominent SS13 codebases, I'll take your word for it.

**Comment by Redict** [comment]
Why not Python?

**Comment by shelltitan** [comment]
I think angelscript allows direct calls from c++, I have never seen it used before so I'm not sure about how easy it is to use. I can speak about godot's gdscript which is fairly easy to use it supports some function calls from c++ if you expose it to the script or gdnative which allows c# or even c++ as scripting language.

**Comment by minexew** [comment]
@shelltitan The problem with many game-focused languages like Angelscript, Gdscript, GML or Pawn is that they're simply not very good languages. That's why they have barely any adoption even in their own niche. They were somewhat relevant in the pre-JIT world, where a well-optimized interpreter (which often imposed limits on dynamic typing) was important. Nowadays? Not really.

Needless to say, low adoption is a poor proposition for a language's future. For an example, look at Vala -- a neat language that solves real issues with C for Gtk app development, but it failed to gain critical mass and now is on life support.

**Comment by PJB3005** [comment]
@Redict Python is:
1. Slow
2. Impossible to sandbox without purely OS-level sandboxing.
3. Generally a pain to bind with I heard.
4. GIL might be a pain.

**Comment by Redict** [comment]
@PJB3005 Python is:
1. Fast. Look at Cython.
2. https://pypi.python.org/pypi/pysandbox/
3. As any other script language. Actually, not such a pain.
4.  > In Cython the GIL exists, but can be released temporarily using a "with" statement

**Comment by PJB3005** [comment]
I can't talk about the other points, but that sandboxing thing is "tested with Python 2.6" and unmaintained.

**Comment by blockspacer** [comment]
cling based C++ as scripting language / hot code reload
Why? Able to run C++ script in runtime or compile it for max speed ( as in example https://github.com/derofim/cling-cmake )
Also: May be used to improve modules compile time 

> **HOT code reload**
> possible approaches:
> 
> store app state
> fix cling undo for files
> https://root-forum.cern.ch/t/loading-unloading-class-as-interpreted-macro-in-cling-multiple-times/32976/2
> 
> execute cling code to change callbacks & variables
> nested cling::Interpreter with multiple cling::MetaProcessor
> IDK how to do it, but you can create child cling::Interpreter
